File: android/app/src/main/java/com/flutcloud/flutlink/data/ApiException.kt
package com.flutcloud.flutlink.data

/**
 * Error surfaced by the FlutCloud/WebDAV API layer. Mirrors the desktop
 * client's serialized `AppError { code, message }`.
 */
open class ApiException(
    override val message: String,
    val code: String = "api_error",
    val statusCode: Int = 0
) : Exception(message)

/** The server does not announce the `flutcloud` capability. */
class FlutCloudAppMissing : ApiException(
    "This server does not run the FlutCloud app.", "flutcloud_app_missing"
)

/** Network / HTTP-level failure (DNS, TLS, connection refused, ...). */
class NetworkException(cause: Throwable) : Exception(cause)

File: ios/FlutLink/Data/ApiException.swift
// TEST — iOS port created by opencode. Not a production build.
import Foundation

enum ApiException: LocalizedError {
    case api(message: String, code: String = "api_error", statusCode: Int = 0)
    case flutCloudAppMissing
    case network(underlying: Error)

    var message: String {
        switch self {
        case .api(let message, _, _):
            return message
        case .flutCloudAppMissing:
            return "This server does not run the FlutCloud app."
        case .network(let underlying):
            return underlying.localizedDescription
        }
    }

    var code: String {
        switch self {
        case .api(_, let code, _):
            return code
        case .flutCloudAppMissing:
            return "flutcloud_app_missing"
        case .network:
            return "network_error"
        }
    }

    var statusCode: Int {
        switch self {
        case .api(_, _, let statusCode):
            return statusCode
        default:
            return 0
        }
    }
}
